[CORE-13829] ct/reconciler: Metrics

wdberkeley opened the pull request:
Metrics added:

- objects_uploaded: Total L1 objects successfully built and uploaded.
  If this goes up, the reconciler is doing something.

- bytes_reconciled: Total bytes of objects produced by reconciliation.
  Basic measure of how much reconciliation goes brrr.

- batches_reconciled: Total record batches processed.
  Basic measure of how much reconciliation goes brrr.

- partitions_reconciled: Counts each time a partition contributes
  batches to an object.
  Useful as a numerator or denominator with other metrics. Can show, for
  example, if reconciliation is dominated by a few busy partitions.

- object_build_failed: Objects that failed to build.

- object_upload_failed: Objects that failed to upload.

- empty_objects_skipped: Objects skipped when no data was available.
  This happens if no partition assigned to the object contributes
  batches. It's useful for measuring load distribution-- if many objects are
  skipped but there is high throughput in bytes/batches, reconciliation is
  concentrated ina few active partitions.

Metrics NOT added:

- Objects accepted by the metastore.
  This is hard to count, because add_objects can succeed, but some or
  all of the objects are not actually accepted into L1 by the metastore.
  Because of the complexity, this metric is not in this change.
## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

wdberkeley committed:
ct/reconciler: Add metrics infrastructure and throughput metrics

This adds a metrics probe for the reconciler and a handful of throughput
metrics.

Metrics added:

- objects_uploaded: Total L1 objects successfully built and uploaded.
  If this goes up, the reconciler is doing something.

- bytes_reconciled: Total bytes of objects produced by reconciliation.
  Basic measure of how much reconciliation goes brrr.

- batches_reconciled: Total record batches processed.
  Basic measure of how much reconciliation goes brrr.

- partitions_reconciled: Counts each time a partition contributes
  batches to an object.
  Useful as a numerator or denominator with other metrics. Can show, for
  example, if reconciliation is dominated by a few busy partitions.

- object_build_failed: Objects that failed to build.

- object_upload_failed: Objects that failed to upload.

- empty_objects_skipped: Objects skipped when no data was available.
  This happens if no partition assigned to the object contributes
  batches. It's useful for measuring load distribution-- if many objects
  are skipped but there is high throughput in bytes/batches,
  reconciliation is concentrated in a few active partitions.

Metrics NOT added:

- Objects accepted by the metastore.
  This is hard to count, because add_objects can succeed, but some or
  all of the objects are not actually accepted into L1 by the metastore.
  Because of the complexity, this metric is not in this change.

wdberkeley committed:
ct/reconciler: Factor out test utils

Pulls out the fake_source and unreliable io and metastore into a test
helper file. It'll be useful to have these separate for the metrics
tests.

dotnwat commented on src/v/cloud_topics/reconciler/reconciler_probe.cc:
```diff
@@ -0,0 +1,63 @@
+          "batches_reconciled",
+          [this] { return _batches_reconciled; },
+          sm::description("Total record batches reconciled")),
+        sm::make_counter(
+          "partitions_reconciled",
+          [this] { return _partitions_reconciled; },
+          sm::description("Counts each time a partition contributed a batch")),
```
A bit unclear what value this provides. Like if it is 100, then did 1 partition contribute 100x or did 100 partitions contribute once?

wdberkeley replied:
It counts any time a partition contributes a batch to an object in reconciliation. So if in one round with 12 partitions, 10 provide data to build 3 objects, this metric increments by 10. The next round, if 3 of 12 provide data to 2 objects, it increments by 3. It's a measure of spread of reconciliation work over partitions. Yeah, it is probably the least useful of the metrics, though.

dotnwat replied:
all good. we can prune this down later

dotnwat commented on src/v/cloud_topics/reconciler/reconciler_probe.cc:
```diff
@@ -0,0 +1,63 @@
+          "bytes_reconciled",
+          [this] { return _bytes_reconciled; },
+          sm::description("Total bytes produced, including metadata")),
+        sm::make_counter(
+          "batches_reconciled",
+          [this] { return _batches_reconciled; },
+          sm::description("Total record batches reconciled")),
```
We should think carefully about which metrics should be per-partition. The aggregations that occur in the reconciler kind of anonymize the partitions. I guess we just need to think through the situations we might be concerned about.

wdberkeley replied:
Right now I'm not planning on adding any partition-level metrics.

rockwotj commented on src/v/cloud_topics/reconciler/tests/reconciler_metrics_test.cc:
```diff
@@ -0,0 +1,200 @@
+namespace {
+
+template<typename T>
+std::optional<T> find_metric_value(std::string_view metric_name) {
```
Since this is cribbed from Wasm and uses some implementation details of seastar, can we factor this into a test helper so if/when seastar changes internals we only have to fix it once?

dotnwat reviewed (dismissed):
no concerns from me. i think at this point it is fine to collect anything/everything. as i noted in a comment, some of the metrics might make more sense as per-ctp metrics. but we can figure that out as we go.

rockwotj commented on src/v/cloud_topics/reconciler/tests/reconciler_metrics_test.cc:
```diff
@@ -0,0 +1,200 @@
+    EXPECT_THAT(get_partitions_reconciled(), Optional(3));
+    EXPECT_THAT(get_batches_reconciled(), Optional(6));
+    EXPECT_THAT(get_bytes_reconciled(), Optional(Gt(bytes_after_first)));
+    // TODO: It shouldn't be a failure to skip an empty object.
```
If you rebase then I think you can distinguish between these cases now

rockwotj commented on src/v/cloud_topics/reconciler/tests/reconciler_metrics_test.cc:
```diff
@@ -0,0 +1,200 @@
+    EXPECT_THAT(get_partitions_reconciled(), Optional(2));
+    EXPECT_THAT(get_batches_reconciled(), Optional(5));
+    EXPECT_THAT(get_bytes_reconciled(), Optional(bytes_after_first));
+    // TODO: It shouldn't be a failure to skip an empty object.
+    EXPECT_THAT(get_objects_failed(), Optional(1));
```
If you rebase then I think you can distinguish between these cases now

wdberkeley replied:
I don't think so... the issue is that `build_and_put_object` returns an error, which we use to put the object in the failed object list so it gets removed from the builder, but there isn't a way to tell why something failed in the failed list. It's fixable, I'm just not sure it's worth fixing other than the deeper fix of not trying to reconcile partitions where LRO == LSO.

rockwotj replied:
I guess if you changed the stat to be `reading from source failed vs succeed` and moved the upload stat down into where we actually do the upload you'd be able to distinguish

wdberkeley replied:
Yeah I split failed uploads and failed builds and then it's easy to tell the difference. Not sure why I didn't do that in the first place 🤷‍♀️ .

rockwotj reviewed (commented):
LGTM

WillemKauf commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -188,6 +189,8 @@ ss::future<> reconciler::reconciliation_loop() {
 }
 
 ss::future<> reconciler::reconcile() {
+    _probe.increment_rounds();
```
This doesn't seem like a particularly useful metric at the moment, given we are polling at a fixed rate. Whenever we land at a different forcing function or multiple (spatial limit, temporal limit) it may be useful to disambiguate "why" a round of reconciliation was triggered through this metric or a set of metrics (e.g reconciliation round triggered by space threshold, triggered by temporal threshold, etc)...

wdberkeley replied:
Yeah, it's a bottom two usefulness of the set, but this kind of dead simple metric is sometimes really useful because it's very cheap and it tells you if _anything_ is happening.

WillemKauf commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -413,6 +421,8 @@ reconciler::build_object(
         auto meta = co_await add_source_to_object(ctx, src);
         if (meta.has_value()) {
+            _probe.increment_partitions_reconciled();
+            _probe.add_batches_reconciled(meta->batch_count);
```
Do we need both `batches_reconciled` and `bytes_reconciled`? Not sure what high level information is conveyed by one but not the other.

wdberkeley replied:
Experience tells me it's helpful to count bytes and "ops" or "units of work" in cases like this. They are usually highly correlated, but anomalies in the byte / batches ratio, meaning either a deviation from that cluster's trend or a deviation from generally how similar (or expected to be similar) workloads look, is a valuable signal for operators.

wdberkeley committed:
test_utils: Factor out metric getter

Yoink the wasm metrics getter into a test util file because the
reconciler metrics test will use it.

wdberkeley commented:
Force push to rebase, resolve conflicts, and add the metrics getter refactor at the beginning.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73377-->
<!--65b7979d-61a6-4090-affe-c2a709d944c2-->
<!--title-->
## Retry command for [Build#73377](https://buildkite.com/redpanda/redpanda/builds/73377)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1aecddff-3c77-43f4-ad4d-efb529ba565f-->
<!--TEST RESULTS-->
<!--56291516-3593-4f01-86b4-8d679be8ed4c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73377">build#73377</a></summary>


| test_class               | test_method            | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                |
|:-------------------------|:-----------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| OffsetRetentionTest      | test_offset_expiration | null                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/73377#0199a13f-a43a-4053-a09a-d4e893700520 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration    |
| RandomNodeOperationsTest | test_node_operations   | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73377#0199a139-3812-456d-a673-5e2493e1eb4e | FLAKY         | 20/21    | upstream reliability is '99.14893617021276'. current run reliability is '95.23809523809523'. drift is 3.91084 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73646">build#73646</a></summary>


| test_class                       | test_method                | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests    | test_replication_basic     | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73646#0199bad6-767f-455a-8439-e81942909564 | FLAKY         | 20/21    | upstream reliability is '99.38650306748467'. current run reliability is '95.23809523809523'. drift is 4.14841 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic        |
| DataMigrationsApiTest            | test_mount_inexistent      | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/73646#0199bad5-1589-401f-beb2-a96e399141bf | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent                 |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null                                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/73646#0199bad6-767b-4b3e-acf9-28e62e9cf2fe | FLAKY         | 17/21    | upstream reliability is '92.27405247813411'. current run reliability is '80.95238095238095'. drift is 11.32167 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<!--/body-->

Lazin commented on src/v/cloud_topics/reconciler/reconciler_probe.h:
```diff
@@ -0,0 +1,55 @@
+    uint64_t _batches_reconciled{0};
+    uint64_t _partitions_reconciled{0};
+    uint64_t _objects_failed{0};
+    uint64_t _empty_objects_skipped{0};
```
maybe it makes sense to add a metric that shows the amount of pending work? wdyt?

Lazin replied:
it's not enough to see that the reconciler is doing progress because it may not be able to keep up with the ingress

Lazin replied:
I think I can add the ability to query the number of bytes which are not reconciled yet to the partition's ctp_stm if everyone agrees that this is useful.

dotnwat replied:
i could see us wanting this value in multiple places. a metric of per-ctp backlog, and then the aggregated value at the reconciler.

wdberkeley replied:
A metric for pending works makes a lot of sense and a metric of max bytes yet-to-be-reconciler (max over attached partitions) would be great. I'm adding some saturation metrics in a follow-up to this one.

wdberkeley commented:
Force push to rebase. The only change is replacing objects_failed with object_build_failed and object_upload_failed, which lets us distinguish between failures and empty objects.

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -352,21 +355,27 @@ reconciler::build_and_put_object(
     }
 
+    _probe.increment_objects_uploaded();
+    _probe.add_bytes_reconciled(obj_meta.object_info.size_bytes);
```
I would expect this to be next to `add_batches_reconciled`?

dotnwat replied:
>     _probe.add_bytes_reconciled(obj_meta.object_info.size_bytes);

>I would expect this to be next to add_batches_reconciled?

yeh. did you mean to do something here like `bytes_uploaded`?

wdberkeley replied:
Having it where it was meant we didn't count the bytes of objects that ended up empty if no partition had data to reconcile. I moved the accounting for bytes and batches closer and excluded the empty case with an if.

dotnwat reviewed (commented):
lgtm pending tyler's question

wdberkeley committed:
ct/reconciler: Move tracking of bytes near tracking of batches
